### DISCUSSION TITLE
Potentially allow concurrent stale read processes with an append writer process?

### DISCUSSION THREAD

**Discussion by Mytherin** [duckdb/duckdb#1635] [General]
See background discussion #1448

**Comment by dforsber** [comment]
Great, seems to be much more fit for purpose on doing this here in Discussions.

When going through the details of Howard Chu's LMDB and reflecting back to MVCC and lock free design in general, I think one relevant question here could be to avoid the "world freeze" in high contention situations by allowing multiple versions to co-exist without limits. I can see similar pattern working in Apache Iceberg as well.

It is easy to go side rails here towards "time travel" as well, but maybe that's another Discussion ;).

That said, would it be a big change in design and implementation if instead of 2 versions, there could be potentially virtually limitless versions. The GC is ofc one question that may have its say, but in practice if all readers are gone from an old version, it could be pruned.

**Reply by Mytherin** [reply]
With another layer of indirection it should be possible to have many versions of the system present in the file. The question then indeed becomes when do you prune versions. This is especially problematic in a single version storage format, since it is not as simple as throwing away no-longer required blocks (after all, you cannot throw away random blocks in the middle of a file). What you will likely end up with is a file with a lot of wasted space and many holes.

A multi-file storage format would simplify this (and most other challenges relating to concurrent access from many different processes) significantly.

**Comment by frankier** [comment]
It would be nice to have this also just as a general convenience.

Not sure if this is the place, being a more serious discussion, but I thought I would add a little -- only just barely tested on BTRFS -- Python workaround to use COW which is:

```
def duckdb_connect_roview_cow(filename, *, config=None, fallback="copy"):
    import duckdb
    from reflink import reflink, ReflinkImpossibleError
    from shutil import copy
    from uuid import uuid4
    from os.path import split as psplit, join as pjoin, splitext

    if config is None:
        config = {}
    path, base = psplit(filename)
    base, ext = splitext(base)
    rnd = uuid4().hex
    dest = pjoin(path, f".{base}.rosnap.{rnd}.{ext}")
    try:
        reflink(filename, dest)
    except (NotImplementedError, ReflinkImpossibleError):
        if fallback == "copy":
            copy(filename, dest)
        elif fallback == "connect":
            dest = filename
        elif fallback == "error":
            raise
        else:
            raise ValueError(f"Invalid fallback option: {fallback}")
    return duckdb.connect(dest, read_only = True, config = config)
```

This is the lamest version which just discards the current WAL. Anyone using this should probably add some cleanup of the created files.
